wp-lazy-loading.php

```diff
@@ -149,23 +145,44 @@ function wp_add_lazy_load_attributes( $content, $context = null ) {
 		$tags[] = 'iframe';
 	}
 
+	// Check if the tag exists in $content.
+	foreach ( $tags as $index => $tag_name ) {
+		if ( false === strpos( $content, "<$tag_name" ) ) {
+			unset( $tags[ $index ] );
+		}
+	}
+
 	if ( empty( $tags ) ) {
 		return $content;
 	}
 
 	return preg_replace_callback(
-		'/<(' . implode( '|', $tags ) . ')(\s)[^>]+>/',
-		function( array $matches ) {
-			if ( ! preg_match( '/\sloading\s*=/', $matches[0] ) ) {
-				$tag   = $matches[1];
-				$space = $matches[2];
-
-				return str_replace( '<' . $tag . $space, '<' . $tag . $space . 'loading="lazy" ', $matches[0] );
+		'/<(' . implode( '|', $tags ) . ')\s[^>]+>/',
+		function( array $matches ) use( $content, $context ) {
+			if ( ! preg_match( '/\bloading\s*=/', $matches[0] ) ) {
+				$tag_name = $matches[1];
+				$old_html = $matches[0];
+
+				$new_html = str_replace( "<$tag_name", "<$tag_name" . ' loading="lazy"', $old_html );
+
+				/**
+				 * Filters the tag HTML after adding the `loading` attribute.
+				 *
+				 * The variable part of the filter name is the filtered tag's name. For example for `img`
+				 * the filter would become `wp_add_lazy_loading_to_img`.
+				 *
+				 * @since (TBD)
+				 *
+				 * @param string $new_html The tag's HTML after adding the attribute.
+				 * @param string $old_html The tag's HTML before adding the attribute.
+				 * @param string $content The HTML content that is being filtered.
+				 * @param string $context Optional. Additional context. Defaults to `current_filter()`.
+				 */
+				return apply_filters( "wp_add_lazy_loading_to_{$tag_name}", $new_html, $old_html, $content, $context );
```

Comment:
I don't like the idea of introducing another filter just for this functionality. What would this be useful for? I think we should try to find a solution that only relies on `wp_lazy_loading_enabled`.

Reply:
This filter allows more granular control and follows the same "architecture logic" as when adding the `srcset` and `sizes` attributes. For example plugins that currently add `skip-lazy` to img tags, can use this filter to either stop adding of `loading="lazy"` or replace it with `loading="eager"` etc.

This will allow much better interoperability and "fine tuning".

Reply:
The filter adds quite a bit of overhead for a very specific case. There already is a filter to manage the addition of a `loading` attribute by core, having two is quite confusing.

I see what you're saying that this gives control on a per `img` basis, but I think it'd be much cleaner to handle this internally, which we already do, by checking whether there already is a `loading` attribute. This is a simple solution that plugin developers can leverage:
* They could simply add a `loading="eager"` so ensure that core skips it. Not only does this get rid of the need for a filter, it also encourages the use of standard attributes to indicate the loading method over arbitrary data attributes or class names.
* Even if for whatever reason a plugin wants to or has to keep using `skip-lazy` classes or similar, it could just automatically add `loading="eager"` if that class is present.

Reply:
> The filter adds quite a bit of overhead for a very specific case.

Yes, there is some overhead but it's almost negligible when running the filter without any callbacks. However if a plugin needs to do something on a per-image basis, it would most likely need to add another `preg_replace_callback()` or similar. 

> I think it'd be much cleaner to handle this internally, which we already do, by checking whether there already is a loading attribute.

Yep, that's the "ideal" case.

> They could simply add a loading="eager" so ensure that core skips it.
> ...
> it could just automatically add loading="eager" if that class is present.

Right, but how are the plugins going to make these changes? As far as I see there are two possibilities in general:
- The `skip-lazy` class name or attribute was added to post_content (i.e. it's hard-coded). To handle the new `loading` attribute the plugin would have to run the content through another regex.
- The `skip-lazy` is being added by some regex filtering of the user content, perhaps a shortcode. In this case it may be possible to change the callback and also add `loading="eager"`, but perhaps not in all cases.

Having a per-image filtering also matches how srcset and sizes are handled: https://core.trac.wordpress.org/browser/tags/5.3.2/src/wp-includes/media.php#L1297 and https://core.trac.wordpress.org/browser/tags/5.3.2/src/wp-includes/media.php#L1407.

Thinking that a per-image filter will see more use at the beginning, then little by little will be used less and less as browser support is being refined and all plugins that currently do lazy-loading adopt the `loading` attribute

Reply:
> The `skip-lazy` class name or attribute was added to post_content (i.e. it's hard-coded). To handle the new `loading` attribute the plugin would have to run the content through another regex.

I think even with this filter a regex may be necessary. For someone using `data-skip-lazy` they might get away with `strpos( $str, ' data-skip-lazy=' )`, but with a `skip-lazy` class you'd already need a regex since there might be other classes.

My main problem with this filter though is that it does not clarify its intent. The name does, but the arguments are just two full `img` tags, one looks different than the other.

> Having a per-image filtering also matches how srcset and sizes are handled: https://core.trac.wordpress.org/browser/tags/5.3.2/src/wp-includes/media.php#L1297 and https://core.trac.wordpress.org/browser/tags/5.3.2/src/wp-includes/media.php#L1407.

Both of these filter specific data, one the sources, the other the sizes, so I don't think these are comparable. Here we're passing through the entire tag, twice. I'd be okay including a more targeted filter like `wp_set_image_loading` which would only filter the value of the `loading` attribute that core would add. We could even allow empty, and then not add it. The signature could be:

`apply_filters( 'wp_set_image_loading', 'lazy', $tag )` (where `$tag` would be `$matches[0]`)

What do you think?

Reply:
Yeah, I see your point. Filtering the whole img tag is more like https://github.com/WordPress/wp-lazy-loading/pull/9 which may eventually happen one day but is out of scope here.

> I'd be okay including a more targeted filter like `wp_set_image_loading` which would only filter the value of the loading attribute that core would add. We could even allow empty, and then not add it.

Yes, makes sense. And yes, in this case it will have to support `false` (or any "falsey" like empty string?) to skip adding the loading attribute. Thinking perhaps it should still pass `$content` and `$context` so plugins have all data necessary to make a decision. More context is always good :)

Patch coming up.

Reply:
Overall sounds good to me. I've added follow-up comments in a review of the updated PR below.